clients/drcachesim/tracer/func_trace.cpp

```diff
@@ -180,16 +196,76 @@ instru_funcs_module_load(void *drcontext, const module_data_t *mod, bool loaded)
     if (drcontext == NULL || mod == NULL)
         return;
 
-    NOTIFY(1, "instru_funcs_module_load, mod->full_path=%s\n",
-           mod->full_path == NULL ? "" : mod->full_path);
-    for (size_t i = 0; i < funcs.entries; i++) {
-        func_metadata_t *f = (func_metadata_t *)drvector_get_entry(&funcs, (uint)i);
+    const char *mod_name = dr_module_preferred_name(mod);
+    if (mod_name == nullptr) {
+        const char *slash = strrchr(mod->full_path, '/');
+#ifdef WINDOWS
+        const char *bslash = strrchr(mod->full_path, '\\');
+        if (bslash != nullptr && bslash > slash)
+            slash = bslash;
+#endif
+        if (slash != nullptr)
+            mod_name = slash + 1;
+    }
+    if (mod_name == nullptr)
+        mod_name = "<unknown>";
+    NOTIFY(2, "instru_funcs_module_load for %s\n", mod_name);
+    for (size_t i = 0; i < func_names.entries; i++) {
+        func_metadata_t *f = (func_metadata_t *)drvector_get_entry(&func_names, (uint)i);
         app_pc f_pc = get_pc_by_symbol(mod, f->name);
         if (f_pc != NULL) {
-            if (drwrap_wrap_ex(f_pc, func_pre_hook, func_post_hook, (void *)i, 0)) {
-                NOTIFY(1, "Inserted hooks for function %s\n", f->name);
+            dr_mutex_lock(funcs_wrapped_lock);
+            int id = wrap_id++;
+            drvector_append(&funcs_wrapped,
+                            create_func_metadata(f->name, id, f->arg_num));
+            char qual[DRMEMTRACE_MAX_QUALIFIED_FUNC_LEN];
+            int len = dr_snprintf(qual, BUFFER_SIZE_ELEMENTS(qual), "%s!%s\n", mod_name,
+                                  f->name);
+            if (len < 0 || len == BUFFER_SIZE_ELEMENTS(qual)) {
+                NOTIFY(0, "Qualified name is too long and was truncated: %s!%s\n",
+                       mod_name, f->name);
+            }
+            NULL_TERMINATE_BUFFER(qual);
+            size_t sz = strlen(qual);
+            if (write_file_func(funclist_fd, qual, sz) != static_cast<ssize_t>(sz))
+                NOTIFY(0, "Failed to write to funclist file\n");
+            dr_mutex_unlock(funcs_wrapped_lock);
+            if (drwrap_wrap_ex(f_pc, func_pre_hook, func_post_hook,
+                               (void *)(ptr_uint_t)id, 0)) {
+                NOTIFY(1, "Inserted hooks for %s!%s == id %d\n", mod_name, f->name, id);
             } else {
-                NOTIFY(1, "Failed to insert hooks for function %s\n", f->name);
+                // We could not write to the file and try to re-use the failed id (or
+                // write "symbol,id" to the file) to indicate this was not traced, but
+                // that seems like more complexity than it's worth: it seems simpler to
+                // list it as something we tried to trace and just have no data in the
+                // trace, counting on this warning to notify the user.
+                //
+                // TODO: The most likely reason to come here is two symbols mapping to
```

Comment:
Yes, this should be very common. Not just for operator new and its many versions, but also aliases from replacements (say malloc() and tc_malloc()).
I think with the current behavior users still have to handle duplicates, except it would be harder for them with only a log message. So "symbol, id" sounds like the way to go.

Reply:
Agreed.  Done.